### ISSUE TITLE
poll_file ignores or hangs on Ctrl-C

### ISSUE THREAD

**Issue by rened** [JuliaLang/julia#9538]
On both OSX and Linux in julia 0.3.3 the following cannot be interrupted with Ctrl-C, and when you hit Ctrl-C a lot of times it might hang altogether (i.e. not return after `timeout==1` seconds):

```jl
poll_file("test",0.1,1)
```

**Comment by ihnorton** [comment]
Reproduced on master.

**Comment by JeffBezanson** [comment]
Reduced test case:

```
begin
  a = Condition()
  b = Condition()
  T = @schedule wait(b)
  wait(a)
end
```

The basic issue is that there is no connection between task `T` and its outer task, which is the one we really want to interrupt. The first ^C kills task `T`, and we end up in `wait()` inside `task_done_hook` for it. From there ^C kills the task "again" and we loop back into `task_done_hook`, rinse and repeat. This is a minimal patch to improve the situation:

```
diff --git a/base/task.jl b/base/task.jl
index f732343..b555ce8 100644
--- a/base/task.jl
+++ b/base/task.jl
@@ -80,7 +80,11 @@ function task_done_hook(t::Task)
         end
         yieldto(nexttask, result)
     else
-        wait()
+        try
+            wait()
+        catch e
+            throwto(nexttask, e)
+        end
     end
 end
```

With that the first ^C kills `T`, and the next ^C kills its next task and you get the prompt back. There are two further refinements to discuss:

1. If the task was killed by an InterruptException, immediately propagate that to its next task rather than entering `wait()` first.
1. If the next task is also done, have a task-of-last-resort available. In the REPL this would be the task that runs the user input. In fact a better option might be to always propagate interrupts to that task.

cc @vtjnash @Keno

**Comment by JeffBezanson** [comment]
Ok this patch seems to work pretty well:

```
diff --git a/base/REPL.jl b/base/REPL.jl
index 42b9a2c..4b17651 100644
--- a/base/REPL.jl
+++ b/base/REPL.jl
@@ -82,7 +82,7 @@ end
 
 function start_repl_backend(repl_channel::RemoteRef, response_channel::RemoteRe
f)
     backend = REPLBackend(repl_channel, response_channel, nothing)
-    @async begin
+    global interactive_task = @schedule begin
         # include looks at this to determine the relative include path
         # nothing means cwd
         while true
diff --git a/base/task.jl b/base/task.jl
index f732343..de883d8 100644
--- a/base/task.jl
+++ b/base/task.jl
@@ -80,6 +80,9 @@ function task_done_hook(t::Task)
         end
         yieldto(nexttask, result)
     else
+        if err && isa(result,InterruptException)
+            throwto(REPL.interactive_task, result)
+        end
         wait()
     end
 end
```

Will have to work out the details of how `interactive_task` gets set.

**Comment by vtjnash** [comment]
I've considered doing that. However, this can result in the task getting added multiple times to the workqueue. I believe the correct solution here is going to be to allow waiting on multiple Tasks, not just Condition objects. For that, I believe there should be a level-triggered object containing an event queue which could be added to the waitq of an edge-triggered Condition object (supporting an async multi-producer/multi-consumer model).

**Comment by JeffBezanson** [comment]
Since we control the REPL, it should be possible to have a ^C that always gives the prompt back. I think we could avoid the multiple-queueing here by checking that the interactive_task is in the `:waiting` state first.

We could also consider always raising the ^C error in the interactive task, rather than in whatever task happens to be running at the time.

Of course we should implement waiting on multiple things, but I don't think that's enough to fix this. In my example above there's no multi-waiting, just the ordinary forever-blocking `wait(Condition())`, plus another task that's also waiting for something unrelated. `poll_file` should use proper multi-waiting, but code like my example will still be possible.

**Comment by vtjnash** [comment]
I agree that it's better to always interrupt the REPL task. That makes the jl_ignore_sigint call make more sense too. 

Multi waiting makes it more likely and explicit that something else is actually waiting for the task, thus giving a path for errors (in general) to get propagated back to the user and reported.
